code/code/spec/spec_mobs_combat.cc

```diff
@@ -10,27 +10,44 @@
 #include "disc_afflictions.h"
 #include "materials.h"
 
-int vampire(TBeing *ch, cmdTypeT cmd, const char *, TMonster *myself, TObj *)
-{
-  if ((cmd != CMD_MOB_COMBAT) || !myself->awake())
-    return FALSE;
+int vampire(TBeing* ch, cmdTypeT cmd, const char*, TMonster*, TObj*) {
+  auto* victim = ch->fight();
 
-  if (ch->spelltask)
-    return FALSE;
+  // Mob spec procs don't get called if their position is < POSITION_STANDING
+  if (cmd != CMD_MOB_COMBAT || !ch || !victim || !ch->sameRoom(*victim) || ch->spelltask)
+    return false;
+
+  act("$n touches $N...", true, ch, nullptr, victim, TO_NOTVICT);
+  act("$n touches you in an attempt to suck away your energy...", true, ch, nullptr, victim,
+    TO_VICT);
 
-  if (ch->fight() && ch->sameRoom(*ch->fight())) {
-    act("$n touches $N!", TRUE, ch, 0, ch->fight(), TO_NOTVICT);
-    act("$n touches you in an attempt to suck away your energy!",
-        TRUE, ch, 0, ch->fight(), TO_VICT);
+  auto level = ch->GetMaxLevel();
 
-    if (!ch->doesKnowSkill(SPELL_ENERGY_DRAIN))
-      ch->setSkillValue(SPELL_ENERGY_DRAIN,120);
+  // No need to call actual spell function, as this is an innate ability. Just calculate damage,
+  // using mob's level as adv_learn value.
+  int dam = ch->getSkillDam(victim, SPELL_ENERGY_DRAIN, level, level);
+  int vit = dice(number(1, level), 4);
 
-    return energyDrain(ch,ch->fight());
+  if (!dam || victim->getImmunity(IMMUNE_DRAIN) >= 100 ||
+      victim->isLucky(levelLuckModifier(level))) {
+    act("Nothing happens.", false, ch, nullptr, nullptr, TO_ROOM);
+  } else {
+    act("$N screams in agony as energy pours from $S body!", false, ch, nullptr, victim,
+      TO_NOTVICT);
+    act("You scream in agony as energy pours from your body!", false, ch, nullptr, victim, TO_VICT);
+    if (ch->reconcileDamage(victim, dam, SPELL_ENERGY_DRAIN) == -1) {
```

Comment:
Just noticed this is a potential crash waiting to happen from a null dereference. Will fix this shortly.

Reply:
Resolved in https://github.com/sneezymud/sneezymud/pull/403/commits/0b09ce281205e09697261c4b31b6055b6c35e057